src/common/Browser.ts

```diff
@@ -539,30 +539,48 @@ export class Browser extends EventEmitter {
    * ```
    */
   async waitForTarget(
-    predicate: (x: Target) => boolean,
+    predicate: (x: Target) => boolean | Promise<boolean>,
     options: WaitForTargetOptions = {}
   ): Promise<Target> {
     const { timeout = 30000 } = options;
-    const existingTarget = this.targets().find(predicate);
+    let remainingTimeout = timeout;
+    const existingTarget = await helper.waitWithTimeout<Target | undefined>(
+      (async () => {
+        const startTime = Date.now();
+        let existingTarget: Target | undefined;
+        for (const target of this.targets()) {
```

Comment:
I am thinking about the following scenario:

1) there is one target existing but it does not match the predicate. The predicate check is async and take 1 sec.
2) while the predicate check is running a new target gets created the matches the predicate.
3) the predicate from step 1 resolves but the local this.targets() does not contain the target created in step 1 because `this.targets()` was called before the target was created.
4) since no existing target is found the function sets up TargetCreated/TargetChanged listeners but they never fire because the target is already created. The entire waitForTarget times out although a matching target is available.

WDYT?

Reply:
I moved the existing targets check to the same time as TargetCreated/TargetChanged.